File: Minimum-Time-Difference/MTD.py
class Solution(object):

    def findMinDifference(self, timePoints):
        """
        :type timePoints: List[str]
        :rtype: int
        """

        minDict = {}  # make dict to store all time by Set
        for s in timePoints:
            cutStr = s.split(":")
            try:
                if int(cutStr[1]) in minDict[cutStr[0]]:
                    return 0  # if this time already in Set, return 0 directly
                else:
                    minDict[cutStr[0]].add(int(cutStr[1]))
            except:
                minDict[cutStr[0]] = {int(cutStr[1])}
        print(minDict)

        tempList = []  # change all time to pure minutes
        for hour in sorted(minDict):
            print(hour)
            count = int(hour)
            tempList += [60 * count + i for i in sorted(list(minDict[hour]))]
        # add first hour to end of list again, because it might be circle time
        firstHour = sorted(minDict)[0]
        tempList += [(24 + int(firstHour)) * 60 +
                     i for i in list(minDict[firstHour])]
        print(tempList)

        min = 1339  # a day have 1440 minutes
        for ind in range(1, len(tempList)):
            if abs(tempList[ind] - tempList[ind - 1]) < min:
                min = abs(tempList[ind] - tempList[ind - 1])
        return min


a = Solution()
print(a.findMinDifference(["23:59", "00:00", "22:34"]))  # 1
print(a.findMinDifference(["22:27", "18:42", "09:57", "09:24", "09:26"]))  # 2

File: Minimum-Time-Difference/MTD.swift
import Foundation

func findMinDifference(_ timePoints:[String]) -> Int{
    var minDict = [Int:Set<Int>]()
    for hour in 0...24 {
        minDict[hour] = Set<Int>()
    }
    
    for time in timePoints {
        let thisTime = time.components(separatedBy:":")
        let (hour, mins) = (Int(thisTime[0])!, Int(thisTime[1])!)

        if (minDict[hour]?.contains(mins))! {
            return 0
        }else{
            minDict[hour]?.insert(mins)
        }
    }
    minDict[24] = minDict[0]
    //print(minDict)

    var small = 1339
    for hour in 0..<24 {
        var tempList = Array<Int>()
        if !minDict[hour]!.isEmpty {
            tempList = Array(minDict[hour]!).sorted()
            if !minDict[hour+1]!.isEmpty {
                tempList.append(minDict[hour+1]!.min()!+60)
            }
        }else {
            continue
        }
        
        for ind in 1..<tempList.count {
            if tempList[ind] - tempList[ind-1] < small{
                small = tempList[ind] - tempList[ind-1]
            }
        }
    }
    
    return small
}

findMinDifference(["23:59", "00:00", "22:34"])
findMinDifference(["22:27", "18:42", "09:57", "09:24", "09:26"])
findMinDifference(["23:59", "00:00", "22:34","22:34"])
